### DISCUSSION TITLE
Mapstruct should respect default values in target classes by default when the source value is null

### DISCUSSION THREAD

**Discussion by kurellajunior** [mapstruct/mapstruct#3685] [Ideas]
Currently (1.6.0) [null-Values from the source will be written by default into the target](https://mapstruct.org/documentation/stable/reference/html/#mapping-result-for-null-properties) by calling the setter or a builder function.
I say this is wrong or at least not intuitive for the use case I see common:

You have a domain model. There you have carefully designed default structures, either by initialization at attribute definition or by providing defaults through lomboks Builder.Default annotation or even via the no-Args-Constructor.
Whenever you do so, I claim you never want those to be set to `null` and run into those annoying NullPointerException down the road.

I believe, that mapstruct should never call a setter or builder method with a null value if not told so explicitly.

Changing this default behaviour should not harm current code a lot:

- not calling a setter for an attribute that is not initialized will lead to a null value ⇒ same behaviour
- not calling a setter for an initialized attribute resetting it to null will enhance the codebase.
- code that expects a default value to be overwritten with `null` probably needs a refactoring anyway.

Minimal example
```java
import lombok.*;
import org.mapstruct.*;

import java.util.ArrayList;
import java.util.List;

@Data
class SimpleSourceItem {
    String text = "text";
}

@Data
class SimpleSource {
    List<SimpleSourceItem> items1;
}


@lombok.Builder
@NoArgsConstructor
@AllArgsConstructor
class SimpleTargetItem {
    String text;
}

@lombok.Builder
@NoArgsConstructor
@AllArgsConstructor
class SimpleTarget {
    @lombok.Builder.Default
    List<SimpleTargetItem> items1 = new ArrayList<>();
}

@Mapper()
public abstract class SimpleItemMapper {
    public abstract SimpleTarget mapItems(SimpleSource source);
}
```
I claim, that this should just work:
```java
new SimpleItemMapperImpl().mapItems(new SimpleSource()).items1 != null;
```

Should that be a thing?

**Comment by filiphr** [comment]
This has been like this for a really long time, changing the defaults would be a bit hard. However, I'm open to the idea if there is enough feedback from the community about this? One thing with this is that it might affect performance a bit (there would be additional null checks in the code)

**Comment by kurellajunior** [comment]
“That is how we have always done it” is a rather weak argument for me :wink: 

I would argue, that changing the defaults would do almost no harm:

- not setting `null` explicitly by calling the setter (or builder method) with a null argument will leave the attribute with the value `null` ⇒ no change in behaviour
- not calling an initialized attribute with an explicit `null` will increase data integrity for those domain models.

Changing the behaviour would only harm situations where you have implemented a sane default value in your domain model (target) AND gave meaning to a null value in your API (source) model with the notion of deleting it.
Whoever designs such models has very likely a lot of specific code in place to care for the inconsistencies and is very likely not using structured approaches like mapstruct…

**Reply by filiphr** [reply]
>I would argue, that changing the defaults would do almost no harm:

What about the potential performance impact?

>Whoever designs such models has very likely a lot of specific code in place to care for the inconsistencies and is very likely not using structured approaches like mapstruct…

MapStruct is not there to judge how people decide to use MapStruct nor how they define their models.

There are other use cases apart from the default one where MapStruct creates a new instance of the target object. e.g. someone might be using an `ObjectFactory` and loading the target object from a database, after which they want to map the data as is (i.e. `null` values should be mapped).

Due to the fact that there is an extremely easy way to always to the null check through `NullValueCheckStrategy`, I am extremely reluctant in modifying the default. I said earlier, the only thing that might change my mind is more feedback from the community that this is something really needed.

What I would like to instead focus on is improved support for things such as `@NonNull` / `@Nullable` (see #1243). Then if your code is using that MapStruct is going to honor it and do a `null` check before invoking the setter.
